File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int size = s.length();
        if(size < 2) return size;
        if (size == 2){
            return (s.charAt(0) == s.charAt(1)) ?  1 : 2;
        }

        int ans = 0, start = 0, end = 0;
        Set<Character> sub = new HashSet<>();
        
        while(size > end){
            if(!sub.contains(s.charAt(end))){
                sub.add(s.charAt(end));
                ans = Math.max(ans, end - start + 1);
                end++;
            }
            else{
                sub.remove(s.charAt(start));
                start++;
            }
        }
        return ans;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
// ds
    int lengthOfLongestSubstring(string s) {
        int size = s.size();
        if(size < 2) return size;
        if (size == 2){
            return (s[0] == s[1]) ?  1 : 2;
        }
        unordered_set<char> sub;
        int ans = 0, start = 0, end = 0;
        while(size > end){
            if(!sub.count(s[end])){
                sub.insert(s[end]);
                ans = max(ans, end - start + 1);
                end++;
            }
            else{
                sub.erase(s[start]);
                start++;
            }
        }
        return ans;
    }
};